src/libhirte/common/config.c

```diff
@@ -0,0 +1,175 @@
+        return hashmap_sip(option_a->name, strlen(option_a->name), seed0, seed1);
+}
+
+int config_initialize() {
+        config = hashmap_new(sizeof(struct config_option), 0, 0, 0, option_hash, option_compare, NULL, NULL);
+        if (config == NULL) {
+                // error during hashmap initialization
+                return ENOMEM;
+        }
+        return 0;
+}
+
+void config_dispose() {
+        hashmap_free(config);
+}
+
+int config_load_complete_cofiguration(
+                const char *default_config_file,
+                const char *custom_config_file,
+                const char *custom_config_directory) {
+        int result = 0;
+
+        if (default_config_file != NULL) {
+                result = config_load_from_file(default_config_file);
+                if (result != 0) {
+                        return result;
+                }
+        }
+
+        if (custom_config_file != NULL) {
+                result = config_load_from_file(custom_config_file);
+                if (result != 0) {
+                        return result;
+                }
+        }
+
+        if (custom_config_directory != NULL) {
+                // TODO: https://github.com/containers/hirte/issues/148
+        }
+
+        result = config_load_from_env();
+
+        return result;
+}
+
+
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wunused-parameter"
+/*
+ * Handle used to store option parsed from a file into a configuration.
+ *
+ * Returns 1 on success, 0 on error as per ini library requirements.
+ */
+static int parsing_handler(void *user, const char *section, const char *name, const char *value) { // NOLINT
+        if (config_set_value(name, value) != 0) {
+                return 0;
+        }
+        return 1;
+}
+#pragma GCC diagnostic pop
+
+int config_load_from_file(const char *config_file) {
+        assert(config_file != NULL);
+        return ini_parse(config_file, parsing_handler, config);
+}
+
+int config_load_from_env() {
+        const char *env_vars[] = { ENV_LOG_LEVEL, ENV_LOG_TARGET, ENV_LOG_IS_QUIET };
+
+        const char *option_names[] = { CFG_LOG_LEVEL, CFG_LOG_TARGET, CFG_LOG_IS_QUIET };
+
+        int length = sizeof(env_vars) / sizeof(env_vars[0]);
+        for (int i = 0; i < length; i++) {
+                char *value = getenv(env_vars[i]);
+                if (value != NULL && strlen(value) > 0) {
+                        // environment variable is defined and has a non-empty value, so store it to the configuration
+                        int result = config_set_value(option_names[i], value);
+                        if (result != 0) {
+                                return result;
+                        }
+                }
+        }
+        return 0;
+}
+
+int config_set_value(const char *option_name, const char *option_value) {
+        char *name_copy = strdup(option_name);
+        if (name_copy == NULL) {
+                return ENOMEM;
```

Comment:
```suggestion
                return -ENOMEM;
```
I think in C its best practice to return negative error codes so the caller just does a `r<0` check and then inspects the code. For `strerror` (to get the string description of the code), it then gets turned into a positive one again.

Reply:
Fixed